index.js

```diff
@@ -22,83 +22,388 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  if (typeof address !== 'number' || address < 0 || address > 0x7f) {
+    throw new Error('Invalid I2C address ' + address
+      + '. Valid addresses are 0 through 0x7f.'
+    );
+  }
+}
+
+function checkRegister(register) {
+  if (register !== undefined &&
+      (typeof register !== 'number' || register < 0 || register > 0xff)) {
+    throw new Error('Invalid I2C register ' + register
+      + '. Valid registers are 0 through 0xff.'
+    );
+  }
+}
+
+function checkLength(length) {
+  if (typeof length !== 'number' || length < 0 || length > 32) {
+    throw new Error('Invalid I2C length ' + length
+      + '. Valid lengths are 0 through 32.'
+    );
+  }
+}
+
+function checkCallback(cb) {
+  if (typeof cb !== 'function') {
+    throw new Error('Invalid I2C callback ' + cb);
+  }
+}
+
+function checkBuffer(buffer) {
+  if (!Buffer.isBuffer(buffer) || buffer.length <= 0 || buffer.length > 32) {
+    throw new Error('Invalid I2C buffer ' + buffer
+      + '. Valid lengths are 0 through 32.'
+    );
+  }
+}
+
+function checkByte(byte) {
+  if (typeof byte !== 'number' || byte < 0 || byte > 0xff) {
+    throw new Error('Invalid I2C byte ' + byte
+      + '. Valid values are 0 through 0xff.'
+    );
+  }
+}
+
+function checkWord(word) {
+  if (typeof word !== 'number' || word < 0 || word > 0xffff) {
+    throw new Error('Invalid I2C word ' + word
+      + '. Valid values are 0 through 0xffff.'
+    );
+  }
+}
+
+var checkAlive = '__r$396836_0$__';
+var devices = '__r$396836_1$__';
+var getDevice = '__r$396836_2$__';
 
 export class I2C extends Peripheral {
-  constructor(baudRate, pins) {
-    super(pins || [ 'SDA0', 'SCL0' ]);
-    if (baudRate) {
-      if (typeof baudRate != number || baudRate % 1000 != 0) {
-        throw new Error('Invalid I2C baud rate. Baud rates must be a multiple of 1000');
-      }
-      sh.run('gpio load i2c ' + baudRate);
-    } else {
-      sh.run('gpio load i2c'); // Is this still necessary?
+  constructor(options) {
+    var pins = options;
+    if (!Array.isArray(pins)) {
+      options = options || {};
+      pins = options.pins || [ 'SDA0', 'SCL0' ];
     }
+    super(pins);
+
+    Object.defineProperties(this, {
+      [devices]: {
+        writable: true,
+        value: []
+      }
+    });
+
+    execSync('modprobe i2c-dev');
   }
 
   destroy() {
+    this[devices].forEach(device => {
+      device.closeSync()
+    });
+
+    this[devices] = [];
+
     super.destroy();
   }
 
+  [checkAlive]() {
+    if (!this.alive) {
+      throw new Error('Attempted to access a destroyed I2C peripheral');
+    }
+  }
+
+  [getDevice](address) {
+    var device = this[devices][address];
+
+    if (device === undefined) {
+      device = i2c.openSync(getBoardRevision() === VERSION_1_MODEL_B_REV_1 ? 0 : 1);
+      this[devices][address] = device;
+    }
+
+    return device;
+  }
+
   // function cb(err, buffer), returns undefined, register is optional
+  // Required by J5
   i2cRead(address, register, length, cb) {
```

Comment:
I cant believe I missed this before, but none of these methods should be prefixed ```i2c``` since they're already members of an I2C class. So this one should be, e.g. ```read(address, register, length, cb)```